HPCC-18969 Implement wizard support for new config manager

kenrowland opened the pull request:
Add support for templated configuration changes

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

Comment out all code, just headers

<!-- Thank you for submitting a pull request to the HPCC project

 PLEASE READ the following before proceeding.

 This project only accepts pull requests related to open JIRA issues.
 If suggesting a new feature or change, please discuss it in a JIRA issue first.
 If fixing a bug, there should be an issue describing it with steps to reproduce.
 The title line of the pull request (and of each commit within it) should refer to the
 associated issue using the format:

 HPCC-nnnnn Short description of issue

 This will allow the Jira ticket to be automatically updated to refer to this pull request,
 and will ensure that the automatically-generated changelog is properly formatted.
 Where a pull request contains a single commit the pull request title will be set automatically,
 assuming that the commit has followed the proper guidelines.

 Please go over all the following points, and put an `x` in all the boxes that apply. You may find
 it easier to press the 'Create' button first then click on the checkboxes to edit the comment.
-->

## Type of change:
- [ ] This change is a bug fix (non-breaking change which fixes an issue).
- [x] This change is a new feature (non-breaking change which adds functionality).
- [ ] This change improves the code (refactor or other change that does not change the functionality)
- [ ] This change fixes warnings (the fix does not alter the functionality or the generated code)
- [ ] This change is a breaking change (fix or feature that will cause existing behavior to change).
- [ ] This change alters the query API (existing queries will have to be recompiled)

## Checklist:
- [x] My code follows the code style of this project.
  - [x] My code does not create any new warnings from compiler, build system, or lint.
- [x] The commit message is properly formatted and free of typos.
  - [x] The commit message title makes sense in a changelog, by itself.
  - [x] The commit is signed.
- [ ] My change requires a change to the documentation.
  - [ ] I have updated the documentation accordingly, or...
  - [ ] I have created a JIRA ticket to update the documentation.
  - [ ] Any new interfaces or exported functions are appropriately commented.
- [x] I have read the CONTRIBUTORS document.
- [x] The change has been fully tested:
  - [x] I have added tests to cover my changes.
  - [x] All new and existing tests passed.
  - [ ] I have checked that this change does not introduce memory leaks.
  - [ ] I have used Valgrind or similar tools to check for potential issues.
- [x] I have given due consideration to all of the following potential concerns:
  - [x] Scalability
  - [x] Performance
  - [ ] Security
  - [ ] Thread-safety
  - [x] Premature optimization
  - [x] Existing deployed queries will not be broken
  - [x] This change fixes the problem, not just the symptom
  - [x] The target branch of this pull request is appropriate for such a change.
- [x] There are no similar instances of the same problem that should be addressed
  - [ ] I have addressed them here
  - [ ] I have raised JIRA issues to address them separately
- [ ] This is a user interface / front-end modification
  - [ ] I have tested my changes in multiple modern browsers
  - [ ] The component(s) render as expected

## Testing:
<!-- Please describe how this change has been tested.-->

<!-- Thank you for taking the time to submit this pull request and to answer all of the above-->

kenrowland commented:
@jpmcmu Please approve. I abandoned the old PR because it was against master. This one is for candidate-7.2.0. The old one was open so long that we moved our active branch. There are no code changes since your review, only changes to get the tests introduced to pass in the smoketest.

HPCCSmoketest commented:
Process of PR-11967, label: HPCC-18969 starts now. Estimated completion time is ~ 0.5 hour(s)

jpmcmu commented:
@kenrowland looks good

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 82a03f5f054e7600d366df619b2ab8a624203c72
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |99 |99 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |810 |810 |0 |
|test (thor) |734 |734 |0 |
|test (roxie) |883 |883 |0 |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   38 sec (00:00:38) |  186 sec (00:03:06) |    0 sec (00:00:00) |    3 sec (00:00:03) |   18 sec (00:00:18) | 1483 sec (00:24:43) |   19 sec (00:00:19) | 1747 sec (00:29:07) |

kenrowland commented:
@richardkchapman Please merge.

richardkchapman commented on .gitmodules:
```diff
@@ -34,3 +34,6 @@
 	url = https://github.com/hpcc-systems/aws-sdk-cpp
+[submodule "configuration/configmgr/RapidJSON"]
+        path = configuration/configmgr/RapidJSON
+        url = https://github.com/Tencent/rapidjson.git
```
If submoduling from github, we prefer to fork and then submodule from our local copy (so that our builds don't break if the original repo is removed, renamed, force-pushed etc.)

richardkchapman replied:
Also need to check the licence. In particular looks like we might need to include some acknowledgement that we use it to our binary releases?

What are we using it for that jptree doesn't already provide?

kenrowland replied:
One thing RapidJSON offers is JSON schema support so that modification templates don't have to be validated as part of parsing.

richardkchapman replied:
Licence has been approved by legal, and I have forked the upstream repo. Can you change your PR so that it uses our fork for the submodule please.

richardkchapman replied:
Also probably need to update our licence.txt to mention that we use this component, as per the terms of their licence.

Commit:
HPCC-18969 Implement wizard support for new config manager

Add support for templated configuration changes

Signed-off-by: Ken Rowland <kenneth.rowland@lexisnexisrisk.com>

Comment out all code, just headers

kenrowland commented:
@richardkchapman Change to .gitmodules done. Please merge if no other issues. Thanks.

HPCCSmoketest commented:
Process of PR-11967, label: HPCC-18969 starts now. Estimated completion time is ~ 0.5 hour(s)

HPCCSmoketest commented:
Automated Smoketest: :white_check_mark:
OS: centos 7.4.1708 (Linux 3.10.0-327.28.3.el7.x86_64)
Sha: 747b791214686bceb6d55bc5e1f35def694bdf4a
Build: success
Build: success
HPCC Start: OK

Unit tests result:

| Test | total | passed | failed | errors | timeout | 
|---|---|---|---|---|---|
|unittest |99 |99 |0 |0 |0 |
|wutoolTest(Dali) |19 |19 |0 |0 |0 |
|wutoolTest(Cassandra) |19 |19 |0 |0 |0 |



Regression test result:

| phase | total | pass | fail | 
|---|---|---|---|
|setup (hthor) |11 |11 |0 |
|setup (thor) |11 |11 |0 |
|setup (roxie) |11 |11 |0 |
|test (hthor) |812 |812 |0 |
|test (thor) |736 |736 |0 |
|test (roxie) |885 |885 |0 |


HPCC Stop: OK
Time stats:

| Prep time | Build time | Package time | Install time | Start time | Test time | Stop time | Summary | 
|---|---|---|---|---|---|---|---|
|   29 sec (00:00:29) |  186 sec (00:03:06) |    0 sec (00:00:00) |    3 sec (00:00:03) |   18 sec (00:00:18) | 1489 sec (00:24:49) |   17 sec (00:00:17) | 1742 sec (00:29:02) |
